File: HCMUTE-CTF_2026/be-careful/brute_vm.py
from pathlib import Path

raw = Path(r"C:\Users\Phuc\Downloads\chall").read_bytes()
prog = raw[0x7b220:0x7b220 + 0x600]
tab = raw[0x7b020:0x7b020 + 0x400]

def key(index, seed):
    value = seed ^ 0xA7
    for i in range(index + 1):
        value = ((value * 73 + i + 0x29) ^ (value >> 3)) & 0xff
    return value

def dec(mem, pos, seed): return mem[pos] ^ key(pos, seed)
def ins(pc): return [dec(prog, pc * 5 + i, 0x46) for i in range(5)]

# The 96 operations before the bit checks are all reversible.
ops = [ins(pc) for pc in range(96)]
checks = [ins(pc) for pc in range(96, 243)]

need = [0] * 21
mask = [0] * 21
for op, a, b, lo, hi in checks:
    assert op == 5
    bit, wanted = (b >> 1) & 7, b & 1
    mask[a] |= 1 << bit
    need[a] |= wanted << bit
print('final masks:', [f'{x:02x}' for x in mask])
print('final low bits:', [f'{x:02x}' for x in need])
free_bit = [next(bit for bit in range(8) if not (mask[i] & (1 << bit))) for i in range(21)]

# Decode lookup pages and build their inverse maps.
pages = []
for page in range(2):
    forward = [dec(tab, page * 256 + x, 0x5b) for x in range(256)]
    inverse = [-1] * 256
    for x, y in enumerate(forward):
        if inverse[y] != -1:
            raise RuntimeError('non-bijective lookup page')
        inverse[y] = x
    pages.append(inverse)

def ror8(x, n):
    n &= 7
    return ((x >> n) | (x << (8 - n))) & 0xff

def reconstruct(bits):
    v = [need[i] | (((bits >> i) & 1) << free_bit[i]) for i in range(21)]
    # Invert cross-byte stage.
    for op, a, b, lo, hi in reversed(ops):
        if op == 8:
            v[a] ^= v[b]
        elif op == 9:
            v[a] = (v[a] - v[b]) & 0xff
        elif op in (1, 2, 3, 4):
            continue
        else:
            raise RuntimeError((op, a, b))
    # Invert individual transforms (rotate, table lookup, xor) in reverse.
    for op, a, b, lo, hi in reversed(ops):
        if op == 4:
            v[a] = ror8(v[a], b)
        elif op == 3:
            v[a] = pages[b][v[a]]
        elif op == 2:
            v[a] ^= b
    # Invert the initial swaps.
    for op, a, b, lo, hi in reversed(ops):
        if op == 1:
            v[a], v[b] = v[b], v[a]
    return bytes(v)

solutions = []
for bits in range(1 << 21):
    body = reconstruct(bits)
    if all(0x20 <= ch <= 0x7e for ch in body):
        solutions.append(body)
        print(b'UTECTF{' + body + b'}')
print('printable solutions:', len(solutions))

File: HCMUTE-CTF_2026/be-careful/brute_vm.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

static uint8_t raw[0xc76c0];
static uint8_t inv[2][256];
static uint8_t code[284][5];

static uint8_t key(unsigned index, uint8_t seed) {
    unsigned v = seed ^ 0xa7;
    for (unsigned i = 0; i <= index; ++i)
        v = ((v * 73 + i + 0x29) ^ (v >> 3)) & 255;
    return (uint8_t)v;
}
static uint8_t dec(uint32_t base, unsigned index, uint8_t seed) {
    return raw[base + index] ^ key(index, seed);
}
static void getins(unsigned pc, uint8_t x[5]) {
    for (unsigned i = 0; i < 5; ++i) x[i] = dec(0x7b700, pc * 5 + i, 0x52);
}
static uint8_t ror8(uint8_t x, unsigned n) {
    n &= 7; return (x >> n) | (x << ((8 - n) & 7));
}
int main(void) {
    FILE *f = fopen("C:\\Users\\Phuc\\Downloads\\chall", "rb");
    if (!f || fread(raw, 1, sizeof raw, f) != sizeof raw) return 2;
    uint8_t need[21] = {0}, mask[21] = {0}, freebit[21];
    uint8_t x[5];
    for (unsigned p = 0; p < 284; ++p)
        for (unsigned i = 0; i < 5; ++i) code[p][i] = dec(0x7b700, p * 5 + i, 0x52);
    for (unsigned p = 111; p < 279; ++p) {
        for (unsigned i = 0; i < 5; ++i) x[i] = code[p][i];
        unsigned bit = (x[2] >> 1) & 7;
        mask[x[1]] |= 1u << bit;
        need[x[1]] |= (x[2] & 1) << bit;
    }
    for (unsigned i = 0; i < 21; ++i)
        for (unsigned b = 0; b < 8; ++b) if (!(mask[i] & (1u << b))) freebit[i] = b;
    for (unsigned page = 0; page < 2; ++page)
        for (unsigned i = 0; i < 256; ++i) inv[page][dec(0x7b020, page * 256 + i, 0x5b)] = i;

    unsigned found = 0;
    for (uint32_t bits = 0; bits < 1; ++bits) {
        uint8_t v[21];
        for (unsigned i = 0; i < 21; ++i) v[i] = need[i];
        for (int p = 110; p >= 83; --p) {
            uint8_t *x = code[p];
            if (x[0] == 8) v[x[1]] ^= v[x[2]];
            if (x[0] == 9) v[x[1]] -= v[x[2]];
        }
        for (int p = 82; p >= 20; --p) {
            uint8_t *x = code[p];
            if (x[0] == 4) v[x[1]] = ror8(v[x[1]], x[2]);
            if (x[0] == 3) v[x[1]] = inv[x[2]][v[x[1]]];
            if (x[0] == 2) v[x[1]] ^= x[2];
        }
        for (int p = 19; p >= 0; --p) {
            uint8_t *x = code[p];
            uint8_t t = v[x[1]]; v[x[1]] = v[x[2]]; v[x[2]] = t;
        }
        int ok = 1;
        for (unsigned i = 0; i < 21; ++i) if (v[i] < 0x20 || v[i] > 0x7e) ok = 0;
        if (ok) {
            printf("UTECTF{"); fwrite(v, 1, 21, stdout); puts("}");
            ++found;
        }
    }
    fprintf(stderr, "found=%u\n", found);
    return 0;
}
